### PR TITLE
feature/create_target

### PR DESCRIPTION
**Pull request by Rodrigo-Miranda-0**
#### Board:
* [Ticket #Ji11ZH7J](https://trello.com/c/Ji11ZH7J/3-3-as-a-user-i-should-be-able-to-create-a-new-target-so-i-can-find-people-with-similar-interests-in-a-specific-area)
---
#### Description:

* Allows for the creation of targets, which are bonded to a user and a topic.
---
#### Notes:
*
---
#### Tasks:
- [x] Create target
---
#### Risk:
*
---
#### Preview:

### CODE REVIEW

File: app/controllers/api/v1/api_controller.rb [hunk 1]
```diff
@@ -3,6 +3,27 @@ module V1
...
+
+      rescue_from ActiveRecord::RecordNotFound,        with: :render_not_found
+      rescue_from ActiveRecord::RecordInvalid,         with: :render_record_invalid
+      rescue_from ActionController::ParameterMissing,  with: :render_parameter_missing
+
+      private
+
+      def render_not_found(exception)
+        logger.info { exception } # for logging
+        render json: { error: I18n.t('api.errors.not_found') }, status: :not_found
```
Comment (jpascual1994) [inline app/controllers/api/v1/api_controller.rb:15]: Have you added the error messages to the translations file?

File: .reek.yml [hunk 2]
```diff
@@ -13,6 +13,7 @@ detectors:
     - ActiveSupport::TestCase
     - Target::Application
 
+
```
Comment (jpascual1994) [inline .reek.yml]: Remove this line 😄

File: app/controllers/api/v1/targets_controller.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+module Api
+  module V1
+    class TargetsController < ApiController
+      def create
+        @target = TargetService.new(target_params, current_user, params[:latitude], params[:longitude]).create
```
Comment (jpascual1994) [inline app/controllers/api/v1/targets_controller.rb:5]: Why are not sending latitude and longitude on target_params?

File: app/services/api/v1/target_service.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    class TargetService
+      def initialize(target_params, current_user, longitude, latitude)
+        @target_params = target_params
+        @current_user = current_user
+        @longitude = longitude
+        @latitude = latitude
+      end
+
+      def create
+        # byebug
```
Comment (jpascual1994) [inline app/services/api/v1/target_service.rb]: 👀

File: app/controllers/api/v1/targets_controller.rb (hunk 3 above)
Comment (Rodrigo-Miranda-0) [reply app/controllers/api/v1/targets_controller.rb:5]: Due to them not being part of the Target table (i only store location, not the separate coordinates) it interferes with the creating process.
Should i add those attributes to the table?

File: spec/models/target_spec.rb [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+#
+#  fk_rails_...  (topic_id => topics.id)
+#  fk_rails_...  (user_id => users.id)
+#
+require 'rails_helper'
+
+describe 'Target', type: :model do
+  describe 'Validations' do
+    subject { build :target }
+    context 'validations' do
```
Comment (enzofab91) [inline spec/models/target_spec.rb]: Remove this context, since it's telling the same as describe.

**Review by enzofab91 — COMMENTED** [review]
Nice!

File: app/models/target.rb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+#  index_targets_on_user_id   (user_id)
+#
+# Foreign Keys
+#
+#  fk_rails_...  (topic_id => topics.id)
+#  fk_rails_...  (user_id => users.id)
+#
+class Target < ApplicationRecord
+  belongs_to :topic, dependent: :destroy
+  belongs_to :user, dependent: :destroy
```
Comment (jpascual1994) [inline app/models/target.rb]: Do you want to destroy Topics and Users when a target is destroyed? I think that the dependent destroy should be on the other side of the relationships

File: app/services/api/v1/target_service.rb (hunk 4 above)
Comment (jpascual1994) [reply app/services/api/v1/target_service.rb]: Please remove this line 😄

File: app/services/api/v1/target_service.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+        @target_params = target_params
+        @current_user = current_user
+        @longitude = longitude
+        @latitude = latitude
+      end
+
+      def create
+        location = ActiveRecord::Point.new(@longitude, @latitude)
+        @current_user.targets.create!(@target_params.merge(location:))
+      end
```
Comment (juanmferreira93) [inline app/services/api/v1/target_service.rb:14]: don't we need to notified matched users when we create a Target?

Comment (juanmferreira93) [reply app/services/api/v1/target_service.rb:14]: I send a request to access your Trello to see the ticket :)

Comment (Rodrigo-Miranda-0) [reply app/services/api/v1/target_service.rb:14]: The notification system is addressed in another [Ticket](https://trello.com/c/DwVWDJIY/8-5-as-a-user-i-should-receive-notifications-when-a-compatible-target-has-been-created-so-we-can-start-getting-connected).

And im not sure where i can find the request to access the Trello, but i believe you can join using [this](https://trello.com/invite/b/gFx4RBmP/ATTI27dde06fb208a480a9906800d0778f73FCCEC717/target-mvd-api-rodrigo-miranda) link. Let me know if it works!

**Review by juanmferreira93 — APPROVED** [review]

**Review by jpascual1994 — APPROVED** [review]

**Pull request merged**